libsnark/zk_proof_systems/ppzksnark/r1cs_se_ppzksnark/r1cs_se_ppzksnark.hpp

```diff
@@ -0,0 +1,467 @@
+#include <libsnark/zk_proof_systems/ppzksnark/r1cs_se_ppzksnark/r1cs_se_ppzksnark_params.hpp>
+
+namespace libsnark {
+
+/******************************** Proving key ********************************/
+
+template<typename ppT>
+class r1cs_se_ppzksnark_proving_key;
+
+template<typename ppT>
+std::ostream& operator<<(std::ostream &out, const r1cs_se_ppzksnark_proving_key<ppT> &pk);
+
+template<typename ppT>
+std::istream& operator>>(std::istream &in, r1cs_se_ppzksnark_proving_key<ppT> &pk);
+
+/**
+ * A proving key for the R1CS SEppzkSNARK.
+ */
+template<typename ppT>
+class r1cs_se_ppzksnark_proving_key {
+public:
+    // G^{gamma * A_i(t)} for 0 <= i <= sap.num_variables()
+    libff::G1_vector<ppT> A_query;
+
+    // H^{gamma * A_i(t)} for 0 <= i <= sap.num_variables()
+    libff::G2_vector<ppT> B_query;
+
+    // G^{gamma^2 * C_i(t) + (alpha + beta) * gamma * A_i(t)}
+    // for sap.num_inputs() + 1 < i <= sap.num_variables()
+    libff::G1_vector<ppT> C_query_1;
+
+    // G^{2 * gamma^2 * Z(t) * A_i(t)} for 0 <= i <= sap.num_variables()
+    libff::G1_vector<ppT> C_query_2;
+
+    // G^{gamma * Z(t)}
+    libff::G1<ppT> G_gamma_Z;
+
+    // H^{gamma * Z(t)}
+    libff::G2<ppT> H_gamma_Z;
+
+    // G^{(alpha + beta) * gamma * Z(t)}
+    libff::G1<ppT> G_ab_gamma_Z;
+
+    // G^{gamma^2 * Z(t)^2}
+    libff::G1<ppT> G_gamma2_Z2;
+
+    // G^{gamma^2 * Z(t) * t^i} for 0 <= i < sap.degree
+    libff::G1_vector<ppT> G_gamma2_Z_t;
+
+    r1cs_se_ppzksnark_constraint_system<ppT> constraint_system;
+
+    r1cs_se_ppzksnark_proving_key() {};
+    r1cs_se_ppzksnark_proving_key<ppT>& operator=(const r1cs_se_ppzksnark_proving_key<ppT> &other) = default;
+    r1cs_se_ppzksnark_proving_key(const r1cs_se_ppzksnark_proving_key<ppT> &other) = default;
+    r1cs_se_ppzksnark_proving_key(r1cs_se_ppzksnark_proving_key<ppT> &&other) = default;
+    r1cs_se_ppzksnark_proving_key(libff::G1_vector<ppT> &&A_query,
+        libff::G2_vector<ppT> &&B_query,
+        libff::G1_vector<ppT> &&C_query_1,
+        libff::G1_vector<ppT> &&C_query_2,
+        libff::G1<ppT> &G_gamma_Z,
+        libff::G2<ppT> &H_gamma_Z,
+        libff::G1<ppT> &G_ab_gamma_Z,
+        libff::G1<ppT> &G_gamma2_Z2,
+        libff::G1_vector<ppT> &&G_gamma2_Z_t,
+        r1cs_se_ppzksnark_constraint_system<ppT> &&constraint_system) :
+        A_query(std::move(A_query)),
+        B_query(std::move(B_query)),
+        C_query_1(std::move(C_query_1)),
+        C_query_2(std::move(C_query_2)),
+        G_gamma_Z(G_gamma_Z),
+        H_gamma_Z(H_gamma_Z),
+        G_ab_gamma_Z(G_ab_gamma_Z),
+        G_gamma2_Z2(G_gamma2_Z2),
+        G_gamma2_Z_t(std::move(G_gamma2_Z_t)),
+        constraint_system(std::move(constraint_system))
+    {};
+
+    size_t G1_size() const
+    {
+        return A_query.size() + C_query_1.size() + C_query_2.size() + 3
+               + G_gamma2_Z_t.size();
+    }
+
+    size_t G2_size() const
+    {
+        return B_query.size();
```

Comment:
You're right. Let me double-check all the other *_size() functions and commit a fix.